js/models/messages.js

```diff
@@ -427,6 +444,7 @@
                         }
                     }
                     message.set({
+                        schemaVersion  : dataMessage.schemaVersion,
```

Comment:
This is a place we have to look out for when adding new properties. Haven’t had time to think about how to make this less likely to be forgotten in the future.

Reply:
I added `Signal.Types.Message.initializeSchema` to `Whisper.Message::initialize` constructor.

Reply:
Hm. Maybe it would make sense to use a different technique to go from `message` to `dataMessage`? What keys _don't_ we want to come through? What did you see when you spent some time with this part of the code?

Reply:
Yes, ideally we’d only use constructors as the sole way to create an instance of a model so there is a single point to determine what data is valid vs not. Right now we have a mixture of `new Whisper.Message(…)` and later various `set` calls which makes it harder to reason about the final state of `Message`. For specifics, I’d have to spend more time with it but at a high-level I see a few problematic patterns:

- We treat _incoming_, _outgoing_, _verified-change_, _group update_, _end session_, etc. messages all the same, i.e. as `Whisper.Message`. I understand we want to render them at the same time and persist them as a union (all properties flat and explicit `type` field) in `messages` store at the storage level, but it’d be nice to make a distinction at the app level. This would allow us to break down `handleDataMessage` into shorter functions, e.g. `handleIncomingMessage`, `handleOutgoingMessage`, `handleVerifiedChangeMessage`, etc. Happy to share how this would be modeled if we had support for union types.
- We interweave pure construction of the data type as well as side-effects such as `Message::save` and `Conversation::save`. It’d be nice to break the two apart so we could test the side-effect free data type construction using unit tests.

> What keys don't we want to come through?

For `Whisper.Message` specifically we seem to be doing fairly well to _pick_ data vs _rejecting_ it. Going through my own database, I found `dataMessage: null` in `type: 'outgoing'` which seems to be astray. Other than that, it’d be simply nice if there was a single `new Whisper.Message(…)` call that showed all the properties we initialize with.